### PR TITLE
doc: reorganize the API reference of Wasm extensions

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

Reorganized the Wasm extensions references, and tried to make it look better and put all the Wasm extensions in a top-level section in the Extensions page. In order to achieve this, this commits 
1. separated out the "Wasm Service" proto from the common configurations to the proto named `service.proto` 
2. moved the common configurations in `wasm.proto` to `common.proto`
3. used `public import` for both `service.proto` and `common.proto` in the existing `wasm.proto` so that this is not a breaking change.

## Before
Screenshot from 2021-03-08 16-34-34
Screenshot from 2021-03-08 16-34-23

## After
Screenshot from 2021-03-29 21-43-25
Screenshot from 2021-03-29 21-43-41

### CODE REVIEW

**Comment by mathetake** [comment]
cc @PiotrSikora @lizan

File: api/envoy/extensions/wasm/v3/service.proto [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// WasmService is configured as a built-in *envoy.wasm_service* :ref:`WasmService
+// <config_wasm_service>` This opaque configuration will be used to create a Wasm Service.
+message WasmService {
+  // General plugin configuration.
+  PluginConfig config = 1;
+
+  // If true, create a single VM rather than creating one VM per worker. Such a singleton can
+  // not be used with filters.
+  bool singleton = 2;
+}
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/service.proto:26]: Isn't this a breaking change?

Comment (mathetake) [reply api/envoy/extensions/wasm/v3/service.proto:26]: yeah, now I see this is a breaking change :disappointed:

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 2]
```diff
@@ -14,8 +14,11 @@ option java_outer_classname = "WasmProto";
 option java_multiple_files = true;
 option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
-// [#protodoc-title: Wasm]
-// [#extension: envoy.bootstrap.wasm]
```
Comment (phlax) [inline api/envoy/extensions/wasm/v3/wasm.proto:18]: not sure if i misunderstand - but i think we dont want to lose the extension annotation

Comment (phlax) [reply api/envoy/extensions/wasm/v3/wasm.proto:18]: oh i see its moved

**Comment by mathetake** [comment]
Is there any workaround for moving message to another proto without breaking change? (I guess there isn't).

Having WasmService definition in a separate file, not with the common configurations, seems necessary for me to better integrate with extension annotations.

**Comment by mathetake** [comment]
OK I see that we can use multiple protodoc-title annotations in a single proto. Will resolve the breaking change.

**Comment by mathetake** [comment]
I tried public imports but it looks like PVG does not support them (validate.pb.h does not include the corresponding validation headers) and results in a breaking change.

**Comment by mathetake** [comment]
According to https://github.com/envoyproxy/envoy/pull/10743#issuecomment-612931049, we don't need to maintain backward compatibility with PGV, so I think I'm going to go with using `import public`. Will do the follow up in a few days.

**Comment by mathetake** [comment]
@phlax @PiotrSikora PTAL! I updated the PR description what I did in this change.

**Comment by phlax** [comment]
rendered docs are here https://storage.googleapis.com/envoy-pr/15360/docs/index.html

File: api/envoy/extensions/wasm/v3/common.proto [hunk 3]
```diff
@@ -0,0 +1,152 @@
+syntax = "proto3";
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/common.proto:1]: `api/envoy/extensions/wasm` is the `WasmService` extension. The common definitions should be moved to `api/envoy/extensions/common/wasm/v3/wasm.proto` instead.

File: api/envoy/extensions/wasm/v3/service.proto [hunk 4]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
```
Comment (PiotrSikora) [inline api/envoy/extensions/wasm/v3/service.proto:1]: This should probably stay in the original file (`api/envoy/extensions/wasm/v3/wasm.proto`) to follow naming of other extensions.

File: api/envoy/extensions/wasm/v3/common.proto (hunk 3 above)
Comment (mathetake) [reply api/envoy/extensions/wasm/v3/common.proto:1]: I found that keeping same proto package name `envoy.extensions.wasm.v3` while moving to `envoy/extensions/common/wasm/v3/wasm.proto` is not possible (looks like the formatter or something else trying to delete the package). IIUC, renaming the package name would result in a breaking change so I'm stuck in a deadlock :disappointed:

**Review by phlax — CHANGES_REQUESTED** [review]
hi @mathetake 

its not immediately clear why but when i search for wasm with this PR

i see the following link in search

wasm-refactor


clicking the first link leads me to https://storage.googleapis.com/envoy-pr/8c8e1ef/docs/api-v3/extensions/wasm/v3/wasm.proto.html?highlight=wasm

when i check the equivalent page in existing docs, i get:

https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/wasm/v3/wasm.proto.html?highlight=wasm

**Comment by mathetake** [comment]
@phlax Thanks for taking a look. Unfortunately I cannot work on this PR anymore for now since the mv of existing proto files seem much more complex than I thought. 

Will come back here when I get more cycles. Closing for now.

**Pull request closed without merging**